src/Symfony/Component/AssetMapper/AssetMapper.php

```diff
@@ -23,50 +24,24 @@
 class AssetMapper implements AssetMapperInterface
 {
     public const MANIFEST_FILE_NAME = 'manifest.json';
-    private const PREDIGESTED_REGEX = '/-([0-9a-zA-Z]{7,128}\.digested)/';
 
     private ?array $manifestData = null;
-    private array $fileContentsCache = [];
-    private array $assetsBeingCreated = [];
-
-    private array $assetsCache = [];
 
     public function __construct(
         private readonly AssetMapperRepository $mapperRepository,
-        private readonly AssetMapperCompiler $compiler,
+        private readonly MappedAssetFactoryInterface $mappedAssetFactory,
         private readonly PublicAssetsPathResolverInterface $assetsPathResolver,
     ) {
     }
 
     public function getAsset(string $logicalPath): ?MappedAsset
     {
-        if (\in_array($logicalPath, $this->assetsBeingCreated, true)) {
-            throw new \RuntimeException(sprintf('Circular reference detected while creating asset for "%s": "%s".', $logicalPath, implode(' -> ', $this->assetsBeingCreated).' -> '.$logicalPath));
-        }
-
-        if (!isset($this->assetsCache[$logicalPath])) {
-            $this->assetsBeingCreated[] = $logicalPath;
-
-            $filePath = $this->mapperRepository->find($logicalPath);
-            if (null === $filePath) {
-                return null;
-            }
-
-            $asset = new MappedAsset($logicalPath);
-            $this->assetsCache[$logicalPath] = $asset;
-            $asset->setSourcePath($filePath);
-
-            $asset->setPublicPathWithoutDigest($this->assetsPathResolver->resolvePublicPath($logicalPath));
-            $publicPath = $this->getPublicPath($logicalPath);
-            $asset->setPublicPath($publicPath);
-            [$digest, $isPredigested] = $this->getDigest($asset);
-            $asset->setDigest($digest, $isPredigested);
-            $asset->setContent($this->calculateContent($asset));
-
-            array_pop($this->assetsBeingCreated);
+        $filePath = $this->mapperRepository->find($logicalPath);
+        if (null === $filePath) {
+            return null;
         }
 
-        return $this->assetsCache[$logicalPath];
+        return $this->mappedAssetFactory->createMappedAsset($logicalPath, $filePath);
```

Comment:
can't this get `$this` as argument to remove the need for the container/service-closure on the compiler?

Reply:
passing `$this` will make this incompatible with decorating the AssetMapper. as it would not pass the decorator.

Not sure it is needed though. This feature relies on decorating the asset factory, not the AssetMapper.

Reply:
Stof is right. If someone decorated the main `AssetMapper`, then the compilers would get the core, non-decorated one. It's... a bit unfortunate, I admit.